### PR TITLE
UVM dependency removal

### PR DESCRIPTION
**Pull request by ldh4**
[WIP] Modifications to make Nalu-wind buildable without Cuda UVM

Pre-requisite work for converting Nalu-wind to be compatible with HIP.
This is to make Nalu wind to build without Cuda UVM

@tasmith4 @alanw0 @jhux2 @ddement

### CODE REVIEW

**Comment by tasmith4** [comment]
After further investigation, @ldh4 and I learned from @alanw0 that tests without `NGP` in the name are not expected to pass in any build.  After running the unit tests with that filter, they all pass on ascicgpu both with and without UVM.

**Comment by alanw0** [comment]
> After further investigation, @ldh4 and I learned from @alanw0 that tests without `NGP` in the name are not expected to pass in any build. After running the unit tests with that filter, they all pass on ascicgpu both with and without UVM.

@tasmith4 @ldh4 More precisely, tests without `NGP` in the name are not expected to pass in a Cuda build. All unit tests are expected to pass in a CPU build. However, I am surprised to hear that all of the NGP unit-tests pass on Cuda without UVM. For example, the `*.NGP*` filter would include at least 3 unit tests that have tpetra usage (MixtureFractionKernelHex8Mesh.NGP_adv_diff_edge_tpetra*). I guess it's possible those tests don't use the graph assembly code, but I would be surprised. (And the one thing I think we're pretty certain about is that the graph assembly code is not using the correct types for at least some of its Views.)

In any case, I'll work with you and Dong Hun tomorrow to help establish where we're at with all this.

**Comment by tasmith4** [comment]
You're right @alanw0, I should have been more precise, thanks for clarifying.  I was also very surprised all the NGP tests pass without UVM (and am willing to be skeptical of the result, although I double-checked my build), we'll have to discuss this more today as you say.

**Comment by alanw0** [comment]
@tasmith4 In particular we're passing those views (that this commit changes to host views) into the setAllIndices method on the tpetra crsgraph. I guess the explanation could be that we're getting those host views from a dual view, and inside setAllIndices tpetra is doing the sync for us and then using the device views. If that's what is happening then great!

**Comment by tasmith4** [comment]
I don't think that's what's happening, so I'm now even more confused by why these tests pass.

**Comment by alanw0** [comment]
I guess I didn't look closely... Now that I look at this commit again, it *doesn't* change the types of the views we're passing into setAllIndices. It is changing the types of other stuff. So we'll just have to work through this today and see what is going on. We'll probably want to put in a (temporary) check in the code, in TpetraLinearSystem.C, for whether uvm is actually on or not.

File: include/SolverAlgorithm.h [hunk 1]
```diff
@@ -65,7 +66,8 @@ struct NGPApplyCoeff
   mutable stk::mesh::NgpField<double> diagField_;
   stk::mesh::NgpField<int> iblankField_;
 
-  CoeffApplier* deviceSumInto_;
+  Teuchos::RCP<CoeffApplier> deviceSumInto_;
```
Comment (PaulMullowney) [inline include/SolverAlgorithm.h:69]: This concerns me. I think we need a generic CoeffApplier definition.

File: src/FixPressureAtNodeAlgorithm.C [hunk 2]
```diff
@@ -113,6 +113,8 @@ FixPressureAtNodeAlgorithm::execute()
       }
     });
   });
+
+  eqSystem_->linsys_->free_coeff_applier(deviceCoeffApplier);
```
Comment (PaulMullowney) [inline src/FixPressureAtNodeAlgorithm.C:117]: Why is this necessary? Were we leaking memory?

File: src/SolverAlgorithm.C [hunk 3]
```diff
@@ -73,6 +73,9 @@ NGPApplyCoeff::NGPApplyCoeff(EquationSystem* eqSystem)
   if (hasOverset_) {
     iblankField_ = nalu_ngp::get_ngp_field<int>(eqSystem->realm_.mesh_info(), "iblank");
   }
+
+  deviceSumInto_ = Teuchos::RCP<CoeffApplier>(eqSystem->linsys_->get_coeff_applier(), CoeffApplierDestructor(), true);
```
Comment (PaulMullowney) [inline src/SolverAlgorithm.C:77]: See comment above on generic CoeffApplier.

File: include/LinearSystem.h [hunk 4]
```diff
@@ -70,6 +70,15 @@ class CoeffApplier
   
 };
 
+struct CoeffApplierDestructor
+{
+  using ptr_t = CoeffApplier;
+
+  void free(ptr_t* coeffApplier) {
+    kokkos_free_on_device(coeffApplier);
```
Comment (PaulMullowney) [inline include/LinearSystem.h:78]: A safety check here might be helpful.

File: src/overset/AssembleOversetDecoupledAlgorithm.C [hunk 5]
```diff
@@ -40,6 +40,8 @@ void AssembleOversetDecoupledAlgorithm::execute()
     fringeNodes.size(), KOKKOS_LAMBDA(const size_t& i) {
       coeffApplier->resetRows(1, &fringeNodes(i), 0, numDof, 1.0, 0.0);
     });
+
+  eqSystem_->linsys_->free_coeff_applier(coeffApplier);
```
Comment (PaulMullowney) [inline src/overset/AssembleOversetDecoupledAlgorithm.C:44]: We're doing this again. We must have been leaking memory?

File: include/SolverAlgorithm.h (hunk 1 above)
Comment (ldh4) [reply include/SolverAlgorithm.h]: This has been reverted. It does look like this was the main culprit for the double-free issue we were encountering.

File: src/FixPressureAtNodeAlgorithm.C (hunk 2 above)
Comment (ldh4) [reply src/FixPressureAtNodeAlgorithm.C:117]: On the TpetraLinearSystem class, we had to change it so that it is no longer holding onto the host/device coeff appliers as its member variables. So, a new coeff applier is constructed on each request. Because of this, we had to introduce a way to properly deallocate coeff appliers when no longer needed. This is one of the changes that we will likely to revisit to find a better way to handle this.
Just as a note, `free_coeff_applier` is an empty function in the `LinearSystem.h`, so this line will not affect `HypreLinearSystem`.

File: src/SolverAlgorithm.C (hunk 3 above)
Comment (ldh4) [reply src/SolverAlgorithm.C]: This line was also reverted.

File: src/overset/AssembleOversetDecoupledAlgorithm.C (hunk 5 above)
Comment (ldh4) [reply src/overset/AssembleOversetDecoupledAlgorithm.C:44]: Same reason as the above.

File: include/LinearSystem.h (hunk 4 above)
Comment (ldh4) [reply include/LinearSystem.h]: Added.

**Comment by ldh4** [comment]
A couple notes on this commit:
- This builds and runs with and without unified memory support (tested only with `cuda` backend for now).
- Building without unified memory currently shows deficiencies in both memory usage and runtime performance. We are still investigating the source of this and will have follow-on tasks to resolve this.
- Building with unified memory does not have this problem and maintains the pervious performance. When tested with `nrel5mw_refined_rcb` problem (~700M nodes) on 90 nodes of `summit`, the build with this patch ran to its completion where as the build from `Nalu-wind` `master` timed out.

**Review by tasmith4 — APPROVED** [review]

**Comment by ldh4** [comment]
Checking numbers from `nrel5mw_refined_rcb` runs again with its termination count reduced by half showed that there isn't any significant difference in runtime and memory usage between this patched version (uvm turned on) and the master, as expected. Runtime had ~5% speedup.
